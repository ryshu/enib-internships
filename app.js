"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const express_1 = __importDefault(require("express"));
const body_parser_1 = __importDefault(require("body-parser"));
const compression_1 = __importDefault(require("compression")); // compresses requests
const lusca_1 = __importDefault(require("lusca"));
const dotenv_1 = __importDefault(require("dotenv"));
const express_flash_1 = __importDefault(require("express-flash"));
const path_1 = __importDefault(require("path"));
const method_override_1 = __importDefault(require("method-override"));
// Load environment variables from .env file, where API keys and passwords are configured
dotenv_1.default.config({ path: '.env' });
// Loggers
const logger_1 = __importDefault(require("./utils/logger"));
const error_1 = require("./utils/error");
const http_status_codes_1 = __importDefault(require("http-status-codes"));
// router and database
const route_1 = __importDefault(require("./configs/setup/route"));
const cas_1 = __importDefault(require("./configs/setup/cas"));
require("./configs/instances/database"); // Only import to setup
require("./configs/setup/database"); // Only import to setup
const express_session_1 = __importDefault(require("express-session"));
const handle_1 = require("./auth/cas/handle");
// Create Express server
const app = express_1.default();
exports.session = express_session_1.default({
    secret: process.env.SESSION_SECRET,
    resave: false,
    saveUninitialized: true,
});
// Express configuration
app.set('port', process.env.INTERNSHIP_ENIB_API_PORT || 3000);
app.use(compression_1.default());
app.use(body_parser_1.default.json({ limit: '1mb' }));
app.use(body_parser_1.default.urlencoded({ extended: true, limit: '1mb' }));
app.use(express_flash_1.default());
app.use(lusca_1.default.xframe('SAMEORIGIN'));
app.use(lusca_1.default.xssProtection(true));
app.use(exports.session);
// CAS Setup
app.get('/logout', cas_1.default.logout); // Logout pass
app.use('/', cas_1.default.bounce, // CAS bounce to connect user if isn't
(req, _res, next) => {
    if (!req.session.info) {
        // If no session info, handle connection
        handle_1.handleConnection(req)
            .then(() => next())
            .catch((e) => next(e));
    }
    else {
        // else, pass
        next();
    }
}, express_1.default.static(path_1.default.join(__dirname, 'public'), { maxAge: 0 }));
app.use(route_1.default);
/**
 * Error handler setup
 */
app.use(method_override_1.default());
app.use((err, _req, res, _next) => {
    if (err instanceof error_1.APIError) {
        return res.status(err.status).json(err);
    }
    else {
        logger_1.default.error(err);
        return res.status(http_status_codes_1.default.INTERNAL_SERVER_ERROR).send(`Internal server error`);
    }
});
exports.default = app;
//# sourceMappingURL=app.js.map