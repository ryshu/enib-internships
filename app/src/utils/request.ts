import axios from 'axios';
import { Message, MessageBox } from 'element-ui';

const service = axios.create({
  baseURL: process.env.VUE_APP_BASE_API, // url = base url + request url
  timeout: 5000,
  // withCredentials: true // send cookies when cross-domain requests
});

// Request interceptors
service.interceptors.request.use(
  config => config,
  error => {
    Promise.reject(error);
  }
);

// Response interceptors
service.interceptors.response.use(
  response => {
    const res = response.data;
    if (res.code && res.code !== 200 && res.code !== 201) {
      Message({
        message: res.message || 'Error',
        type: 'error',
        duration: 5 * 1000,
      });
      if (res.code === 500) {
        MessageBox.confirm(
          '你已被登出，可以取消继续留在该页面，或者重新登录',
          '确定登出',
          {
            confirmButtonText: '重新登录',
            cancelButtonText: '取消',
            type: 'warning',
          }
        ).then(() => {
          window.location.assign('/logout');
        });
      }
      return Promise.reject(new Error(res.message || 'Error'));
    } else {
      return response.data;
    }
  },
  error => {
    console.log(error);
    Message({
      message: error.message,
      type: 'error',
      duration: 5 * 1000,
    });
    return Promise.reject(error);
  }
);

export default service;
