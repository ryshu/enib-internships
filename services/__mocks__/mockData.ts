export function defaultBusiness() {
    const VALID_BUSINESS: IBusinessEntity = {
        name: 'test',
        country: 'France',
        city: 'Brest',
        postalCode: '29200',
        address: 'TEST',
    };
    return VALID_BUSINESS;
}

export function defaultFiles() {
    const VALID_FILE: IFileEntity = {
        name: 'fichier',
        size: 150,
        type: 'john.doe@enib.fr',
        path: 'S10',
    };
    return VALID_FILE;
}

export function defaultInternships() {
    const VALID_INTERNSHIP: IInternshipEntity = {
        subject: 'test',
        description: 'Stage',
        country: 'France',
        city: 'Brest',
        postalCode: '29280',
        address: 'TEST',
        isLanguageCourse: true,
        isValidated: false,
    };
    return VALID_INTERNSHIP;
}

export function defaultMentoringPropositions() {
    const VALID_MENTORING_PROPOSITION: IMentoringPropositionEntity = {
        comment: 'test',
    };
    return VALID_MENTORING_PROPOSITION;
}

export function defaultStudents() {
    const VALID_STUDENT: IStudentEntity = {
        firstName: 'John',
        lastName: 'Doe',
        email: 'john.doe@enib.fr',
        semester: 'S10',
    };
    return VALID_STUDENT;
}

export function defaultInternshipTypes() {
    const VALID_INTERNSHIP_TYPES: IInternshipTypeEntity = {
        label: 'test',
    };
    return VALID_INTERNSHIP_TYPES;
}
