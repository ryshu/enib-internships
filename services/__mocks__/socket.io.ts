import SubSocket from './subsocket';

export class SocketServer {
    public of(_nsp: string | RegExp): any {
        return new SubSocket();
    }

    public close(fn?: () => void): void {
        fn();
    }

    public on(_event: 'connection', _listener: (socket: any) => void): any {
        return;
    }

    public emit(_event: string, ..._args: any[]): any {
        return jest.fn();
    }

    public to() {
        return this;
    }
}

export default function() {
    return new SocketServer();
}
