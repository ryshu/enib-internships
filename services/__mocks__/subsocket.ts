export default class SubSocket {
    public emit ( ..._args: any[]): void {
        return;
    }
}
