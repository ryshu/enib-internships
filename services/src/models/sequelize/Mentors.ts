import * as Sequelize from 'sequelize';

import database from '../../configs/instances/database';
import Campaigns from './Campaigns';
import MPs from './MentoringPropositions';
import Internships from './Internships';

import { IInternshipEntity, IMentoringPropositionEntity } from '../../declarations';

class Mentors extends Sequelize.Model {
    public static associations: {
        propositions: Sequelize.Association<Mentors, MPs>;
        internships: Sequelize.Association<Mentors, Internships>;
        campaigns: Sequelize.Association<Mentors, Campaigns>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public firstName: string;
    public lastName: string;
    public email: string;
    public role: 'default' | 'admin';

    // Campaigns
    public getCampaigns: Sequelize.BelongsToManyGetAssociationsMixin<Campaigns>;
    public addCampaign: Sequelize.BelongsToManyAddAssociationMixin<Campaigns, Campaigns['id']>;
    public hasCampaign: Sequelize.BelongsToManyHasAssociationMixin<Campaigns, Campaigns['id']>;
    public countCampaigns: Sequelize.BelongsToManyCountAssociationsMixin;

    // Propositions
    public getPropositions: Sequelize.HasManyGetAssociationsMixin<MPs>;
    public addProposition: Sequelize.HasManyAddAssociationMixin<MPs, MPs['id']>;
    public createProposition: Sequelize.HasManyCreateAssociationMixin<IMentoringPropositionEntity>;
    public hasProposition: Sequelize.HasManyHasAssociationMixin<MPs, MPs['id']>;
    public countPropositions: Sequelize.HasManyCountAssociationsMixin;

    // Internships
    public getInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addInternship: Sequelize.HasManyAddAssociationMixin<Internships, Internships['id']>;
    public createInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasInternship: Sequelize.HasManyHasAssociationMixin<Internships, Internships['id']>;
    public countInternships: Sequelize.HasManyCountAssociationsMixin;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    public readonly campaigns?: Campaigns[];
    public readonly internships?: Internships[];
    public readonly propositions?: MPs[];
}

Mentors.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        firstName: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        lastName: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        email: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        role: {
            type: new Sequelize.DataTypes.STRING(40),
            allowNull: true,
            defaultValue: 'default',
        },
    },
    {
        tableName: 'mentors',
        sequelize: database,
    },
);

export default Mentors;
