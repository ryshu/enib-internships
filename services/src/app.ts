import express, { Request, Response, NextFunction } from 'express';
import bodyParser from 'body-parser';
import cors from 'cors';
import compression from 'compression'; // compresses requests
import lusca from 'lusca';
import dotenv from 'dotenv';
import flash from 'express-flash';
import path from 'path';
import methodOverride from 'method-override';

// Load environment variables from .env file, where API keys and passwords are configured
dotenv.config({ path: '.env' });

// Loggers
import logger from './utils/logger';
import { APIError } from './utils/error';
import httpStatus from 'http-status-codes';

// router and database
import router from './configs/setup/route';
import './configs/instances/database'; // Only import to setup
import './configs/setup/database'; // Only import to setup

// Create Express server
const app = express();

// Express configuration
app.set('port', process.env.INTERNSHIP_ENIB_API_PORT || 3000);
app.use(compression());
app.use(bodyParser.json({ limit: '1mb' }));
app.use(bodyParser.urlencoded({ extended: true, limit: '1mb' }));
app.use(flash());
app.use(lusca.xframe('SAMEORIGIN'));
app.use(lusca.xssProtection(true));

app.use(express.static(path.join(__dirname, 'public'), { maxAge: 31557600000 }));

const corsOptions = {
    origin: 'http://localhost:4200',
    optionsSuccessStatus: 200,
};
app.use(cors(corsOptions));

app.use(router);

/**
 * Error handler setup
 */
app.use(methodOverride());
app.use((err: APIError | Error, _req: Request, res: Response, _next: NextFunction) => {
    if (err instanceof APIError) {
        return res.status(err.status).json(err);
    } else {
        logger.error(err);
        return res.status(httpStatus.INTERNAL_SERVER_ERROR).send(`Internal server error`);
    }
});

export default app;
