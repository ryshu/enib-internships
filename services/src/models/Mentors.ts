import * as Sequelize from 'sequelize';

import database from '../configs/instances/database';
import Campaigns from './Campaigns';

class Mentors extends Sequelize.Model implements IMentorEntity {
    public static associations: {
        campaigns: Sequelize.Association<Mentors, Campaigns>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public firstName: string;
    public lastName: string;
    public email: string;
    public role: 'default' | 'admin';

    // Campaigns
    public getCampaigns: Sequelize.BelongsToManyGetAssociationsMixin<Campaigns>;
    public addCampaign: Sequelize.BelongsToManyAddAssociationMixin<Campaigns, Campaigns['id']>;
    public hasCampaign: Sequelize.BelongsToManyHasAssociationMixin<Campaigns, Campaigns['id']>;
    public countCampaigns: Sequelize.BelongsToManyCountAssociationsMixin;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    public readonly campaigns?: Campaigns[];
}

Mentors.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        firstName: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        lastName: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        email: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        role: {
            type: new Sequelize.DataTypes.STRING(40),
            allowNull: true,
            defaultValue: 'default',
        },
    },
    {
        tableName: 'mentors',
        sequelize: database,
    },
);

export default Mentors;
