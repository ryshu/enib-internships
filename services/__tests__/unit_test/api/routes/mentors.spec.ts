import request from 'supertest';

// Include app for super agent, not server
import app from '../../../../src/app';

// Include database connection
import dbSetup from '../../../../src/configs/setup/database';

// Import model for pre-operation before asserting API methods
import Mentors from '../../../../src/models/Mentors';

import { defaultMentors } from '../../../../__mocks__/mockData';

beforeAll((done) => {
    dbSetup.then(() => done()).catch((e) => done(e));
});

describe('GET /mentors', () => {
    beforeEach(async () => {
        // Remove all
        await Mentors.destroy({ where: {} });
    });

    it('NoFile_204', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentors`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentors?limit=Nan`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('Mentors_200', async () => {
        const VALID_MENTOR = defaultMentors();

        await Mentors.create(VALID_MENTOR);
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentors`,
        );
        expect(RESPONSE.status).toBe(200);
        expect(Array.isArray(RESPONSE.body.data)).toBeTruthy();
        expect(RESPONSE.body).toMatchSnapshot({
            data: [
                {
                    createdAt: expect.any(String),
                    updatedAt: expect.any(String),
                },
            ],
        });
    });
});

describe('POST /mentors', () => {
    it('NoBody_400', async () => {
        const RESPONSE = await request(app).post(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentors`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('ValidMentors_200', async () => {
        const VALID_MENTOR = defaultMentors();

        const RESPONSE = await request(app)
            .post(`/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentors`)
            .send(VALID_MENTOR);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });
});

describe('GET /mentors/:id', () => {
    beforeEach(async () => {
        // Remove all
        await Mentors.destroy({ where: {} });
    });

    it('NoFile_204', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentors/10`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentors/{falseEncoding}`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('Mentors_200', async () => {
        const VALID_MENTOR = defaultMentors();

        const CREATED = await Mentors.create(VALID_MENTOR);
        const RESPONSE = await request(app).get(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentors/${CREATED.id}`,
        );
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });
});

describe('PUT /mentors/:id', () => {
    beforeEach(async () => {
        // Remove all
        await Mentors.destroy({ where: {} });
    });

    it('NoFile_204', async () => {
        const RESPONSE = await request(app).put(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentors/10`,
        );
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).put(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentors/{falseEncoding}`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('Mentors_200_UpdateAllData', async () => {
        const VALID_MENTOR = defaultMentors();

        const CREATED = await Mentors.create(VALID_MENTOR);

        // Change some data on VALID_MENTOR
        VALID_MENTOR.firstName = '250';

        const RESPONSE = await request(app)
            .put(`/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentors/${CREATED.id}`)
            .send(VALID_MENTOR);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });

    it('Mentors_200_NotAnyDataUpdated', async () => {
        const VALID_MENTOR = defaultMentors();

        const CREATED = await Mentors.create(VALID_MENTOR);

        const RESPONSE = await request(app)
            .put(`/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentors/${CREATED.id}`)
            .send({});
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });
});

describe('DELETE /mentors/:id', () => {
    beforeEach(async () => {
        // Remove all
        await Mentors.destroy({ where: {} });
    });

    it('NoFile_200', async () => {
        const RESPONSE = await request(app).delete(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentors/10`,
        );
        expect(RESPONSE.status).toBe(200);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).delete(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentors/{falseEncoding}`,
        );
        expect(RESPONSE.status).toBe(400);
    });

    it('Mentors_200', async () => {
        const VALID_MENTOR = defaultMentors();

        const CREATED = await Mentors.create(VALID_MENTOR);

        const RESPONSE = await request(app).delete(
            `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}/mentors/${CREATED.id}`,
        );
        expect(RESPONSE.status).toBe(200);
    });
});
