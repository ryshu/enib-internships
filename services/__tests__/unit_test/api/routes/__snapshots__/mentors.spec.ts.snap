// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /mentors Mentors_200 1`] = `
Object {
  "data": Array [
    Object {
      "createdAt": Any<String>,
      "email": "john.doe@enib.fr",
      "firstName": "John",
      "id": Any<Number>,
      "lastName": "Doe",
      "role": "default",
      "updatedAt": Any<String>,
    },
    Object {
      "createdAt": Any<String>,
      "email": "cas.dev@enib.fr",
      "firstName": "Cas",
      "id": Any<Number>,
      "lastName": "DEV",
      "role": "default",
      "updatedAt": Any<String>,
    },
  ],
  "length": 2,
  "max": 2,
  "page": 1,
}
`;

exports[`GET /mentors/:id Mentors_200 1`] = `
Object {
  "campaigns": Array [],
  "createdAt": Any<String>,
  "email": "john.doe@enib.fr",
  "firstName": "John",
  "id": Any<Number>,
  "internships": Array [],
  "lastName": "Doe",
  "propositions": Array [],
  "role": "default",
  "updatedAt": Any<String>,
}
`;

exports[`POST /mentors ValidMentors_200 1`] = `
Object {
  "createdAt": Any<String>,
  "email": "john.doe@enib.fr",
  "firstName": "John",
  "id": Any<Number>,
  "lastName": "Doe",
  "role": "default",
  "updatedAt": Any<String>,
}
`;

exports[`PUT /mentors/:id Mentors_200_NotAnyDataUpdated 1`] = `
Object {
  "createdAt": Any<String>,
  "email": "john.doe@enib.fr",
  "firstName": "John",
  "id": Any<Number>,
  "lastName": "Doe",
  "role": "default",
  "updatedAt": Any<String>,
}
`;

exports[`PUT /mentors/:id Mentors_200_UpdateAllData 1`] = `
Object {
  "createdAt": Any<String>,
  "email": "john.doe@enib.fr",
  "firstName": "250",
  "id": Any<Number>,
  "lastName": "Doe",
  "role": "default",
  "updatedAt": Any<String>,
}
`;
